src/mappings/entities/mapCorporateActions.ts

```diff
@@ -1,124 +1,105 @@
-      return getTextValue(params[1].get('local_id'));
-    }
-  }
-  throw new Error("Event didn't have a CaID local_id parameter within Ca_Id");
-};
+import { Distribution, DistributionPayment } from '../../types';
+import { getBigIntValue, getCaIdValue, getDistributionValue, getTextValue } from '../util';
+import { EventIdEnum, HandlerArgs, ModuleIdEnum } from './common';
 
 /**
- * Subscribes to the CapitalDistribution (BenefitedClaimed and Reclaimed Event)
+ * Subscribes to the CapitalDistribution events
  */
-export async function mapCorporateActions(
-  blockId: string,
-  eventId: EventIdEnum,
-  moduleId: ModuleIdEnum,
-  params: Codec[],
-  event: SubstrateEvent
-): Promise<void> {
+export async function mapCorporateActions({
+  blockId,
+  eventId,
+  moduleId,
+  params,
+  event,
+}: HandlerArgs): Promise<void> {
   if (moduleId === ModuleIdEnum.Capitaldistribution) {
+    if (eventId === EventIdEnum.Created) {
+      await handleDistributionCreated(blockId, params);
+    }
     if (eventId === EventIdEnum.BenefitClaimed) {
-      await handleHistoryOfPaymentEventsForCA(blockId, eventId, params, event);
-      await handleWithholdingTaxesOfCA(eventId, params, event);
+      await handleBenefitClaimed(blockId, eventId, params, event);
     }
     if (eventId === EventIdEnum.Reclaimed) {
-      await handleHistoryOfPaymentEventsForCA(blockId, eventId, params, event);
+      await handleReclaimed(blockId, eventId, params, event);
     }
   }
 }
 
-/**
- * Handles HistoryOfPaymentEventsForCA entity
- */
-async function handleHistoryOfPaymentEventsForCA(
+const handleDistributionCreated = async (blockId: string, params: Codec[]): Promise<void> => {
+  const [rawDid, rawCaId, rawDistribution] = params;
+
+  const { localId, assetId } = getCaIdValue(rawCaId);
+  const distributionDetails = getDistributionValue(rawDistribution);
+
+  await Distribution.create({
+    id: `${assetId}/${localId}`,
+    identityId: getTextValue(rawDid),
+    localId,
+    assetId,
+    ...distributionDetails,
+    taxes: BigInt(0),
+    createdBlockId: blockId,
+    updatedBlockId: blockId,
+  }).save();
+};
+
+const handleBenefitClaimed = async (
   blockId: string,
   eventId: EventIdEnum,
   params: Codec[],
   event: SubstrateEvent
-) {
-  const eventDid = getTextValue(params[0]);
-  const ticker = await getTickerFromCaId(params, eventId);
-  const localId = await getLocalIdFromCaId(params, eventId);
-  const balance = await getBalanceForCa(params, eventId);
-  const tax = getBigIntValue(params[5]);
-  await HistoryOfPaymentEventsForCa.create({
+): Promise<void> => {
+  const [rawEventDid, , rawCaId, , rawAmount, rawTax] = params;
+
+  const targetId = getTextValue(rawEventDid);
+  const { localId, assetId } = getCaIdValue(rawCaId);
+  const amount = getBigIntValue(rawAmount);
+  const tax = getBigIntValue(rawTax);
+
+  const distribution = await Distribution.get(`${assetId}/${localId}`);
+  distribution.taxes += (amount * tax) / BigInt(1000000);
```

Comment:
Dividing like this can loose accuracy in the number, especially if there are a lot of small transactions. I would think its better to leave it in the smallest unit.

It might be the case that they do get rounded and this is correct though. If thats the case then I think `BigInt(1000000)` be a constant instead of a magic number.

Reply:
I wasn't sure either, then I referred back the tooling-gql and apparently it was done there. I thought it's better to leave that way. I don't have problem removing this as it could easily be handled on the SDK side

Reply:
I think it would be better to not lose accuracy in the DB, and convert it in the SDK.

I think tooling is different since it would be storing the raw events and wouldn't compound the error over many events.